### PR TITLE
Add Open Search Provider

### PR DESCRIPTION
**Pull request by cjames23**
This provider would be a pre-requisite for adding functionality to the Amazon provider for the AWS managed service for OpenSearch. This was discussed in a different [PR](https://github.com/apache/airflow/pull/34693) which will be blocked by this PR now. both PRs would be pre-work for completing https://github.com/apache/airflow/issues/33619 and adding Open Search log integration. 

### Open questions to more experienced contributors:

1. For system test DAG to succeed would an OpenSearch domain need to be setup and if so what is the approach for standing up such infrastructure for integration tests?
2. Any advice on mypy for some of the operator parameters? The goal is to allow users to either pass in the arguments directly or by passing in a class in the case of Search which works but in the case of adding a document the end user of the Operator needs to subclass the Document class. I am not sure what the best approach is for the typing for that parameter. 

Email sent for Lazy Consensus to the dev-list. 

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/opensearch/provider.yaml
Comment (eladkal) [inline airflow/providers/opensearch/provider.yaml]: lets set 2.5.0
all provider releases after 19th of September 2023 will have 2.5.0
https://github.com/apache/airflow/blob/main/PROVIDERS.rst#id5

File: generated/provider_dependencies.json
Comment (eladkal) [inline generated/provider_dependencies.json]: I wonder where did this come from?
this file is auto generated by pre-commits.
maybe you used another version of provider.yaml when you run the tests?

**Comment by vincbeck** [comment]
> For system test DAG to succeed would an OpenSearch domain need to be setup and if so what is the approach for standing up such infrastructure for integration tests?

System tests are not triggered as part of the CI. They are just used as example and also as system test/integration test if anybody wants to execute it (and as such, needs to setup the infra). The approach taken by the community is your write system tests in the Airflow repository but you execute them on your own with your infra set up. As an example, in the Amazon provider package, all system tests rely on AWS services (thus AWS account). These system tests are not triggered as part of Airflow CI but AWS team trigger frequently and results are accessible through a [dashboard](https://aws-mwaa.github.io/open-source/system-tests/dashboard.html).

File: airflow/providers/opensearch/provider.yaml
Comment (cjames23) [reply airflow/providers/opensearch/provider.yaml]: If I do that then it fails the CI check for Airflow 2.4

File: generated/provider_dependencies.json
Comment (cjames23) [reply generated/provider_dependencies.json]: Good question. Let me check through things the only dependency defined that I removed was OpenSearch-dsl because after 2.1.0 of OpenSearch-py that was added in directly to have a single package.

File: airflow/providers/opensearch/provider.yaml
Comment (cjames23) [reply airflow/providers/opensearch/provider.yaml]: Hmm but it is still failing that check. Okay I’ll move that back to 2.5.0

Comment (eladkal) [reply airflow/providers/opensearch/provider.yaml]: That is OK. This PR is depended on https://github.com/apache/airflow/pull/34728

Just set 2.5.0 and when we merge https://github.com/apache/airflow/pull/34728 it will fix the issue

**Comment by cjames23** [comment]
@vincbeck Thank you for the clarification on the example dags in the systems tests.

**Comment by cjames23** [comment]
For question number 2 on the PR when originally cut, the only path forward it seems with mypy was to use the Any type. Still open to other ideas if there is a way to be more specific with the type for the Operators that take high level objects for the high level client implementation of OpenSearch.

File: airflow/providers/opensearch/hooks/opensearch.py
Comment (ferruzzi) [inline airflow/providers/opensearch/hooks/opensearch.py]: Perhaps consider making conn a cached_property?  I know the Amazon provider caches it, I don't know if it's appropriate here or not.  I'll leave it up to you.

File: airflow/providers/opensearch/hooks/opensearch.py
Comment (ferruzzi) [inline airflow/providers/opensearch/hooks/opensearch.py]: [Here and elsewhere] This is a very minor non-blocking nitpick, you can ignore if you wish.  One of my side projects I've been chipping away at is enabling more pydocstyle checks to help standardize our codebase.  One of the ones I am working on enabling enforces the rule that the first line in every docstring needs to be phrased as an imperative.  For example:

```suggestion
    Provide a thin wrapper around the OpenSearch client.
```

Basically, phrase it as if you are telling the method or class what to do, not explaining it to someone else.   If you could make the adjustments, I'd appreciate it.  If not, I can go back and do them when I get to it.  I won't block merging for this.

File: airflow/providers/opensearch/hooks/opensearch.py
Comment (ferruzzi) [inline airflow/providers/opensearch/hooks/opensearch.py]: Is this correct or should it be `OpenSearch`?

File: airflow/providers/opensearch/operators/opensearch.py
Comment (ferruzzi) [inline airflow/providers/opensearch/operators/opensearch.py]:
```suggestion
    Add a new document to a given Index or overwrite an existing one.
```

File: airflow/providers/opensearch/operators/opensearch.py
Comment (ferruzzi) [inline airflow/providers/opensearch/operators/opensearch.py]: [Here and below] I'd prefer to see a more narrow catch here.  Catching `Exception` is generally considered poor form, in my experience.

File: docs/apache-airflow-providers-opensearch/connections/opensearch.rst
Comment (ferruzzi) [inline docs/apache-airflow-providers-opensearch/connections/opensearch.rst]: Should this be OpenSearch? Or elsewhere you used `opensearch connection` so maybe that is the right use here?  Not sure.  Branding and naming issues are always tricky, I'm not sure how the OS project handles their name in use like this.

File: docs/apache-airflow-providers-opensearch/connections/opensearch.rst
Comment (ferruzzi) [inline docs/apache-airflow-providers-opensearch/connections/opensearch.rst]: [Optional] Consider including the default values in the docs here.

File: docs/apache-airflow-providers-opensearch/operators/opensearch.rst
Comment (ferruzzi) [inline docs/apache-airflow-providers-opensearch/operators/opensearch.rst]: Check the generated output but I think when this gets parsed into an actual doc file this is going to read awkwardly.  In the AWS package we say "use :class:`~airflow.providers.opensearch.operators.opensearch.OpenSearchAddDocumentOperator` ..." because "use the..." sounded a little odd when the link got parsed.   I'll leave the call up to you, just a suggestion.

**Review by ferruzzi — COMMENTED** [review]
Have not gotten to the tests yet, left a number of small suggestions and requests but nothing major.  Overall looks good.

File: airflow/providers/opensearch/hooks/opensearch.py
Comment (cjames23) [reply airflow/providers/opensearch/hooks/opensearch.py]: No it should be OpenSearch I missed changing this back once I added opensearch and OpenSearch to the spelling list to avoid spelling linting errors.

File: airflow/providers/opensearch/operators/opensearch.py
Comment (cjames23) [reply airflow/providers/opensearch/operators/opensearch.py]: Fixed to use the base  OpenSearchException for the OpenSearchClient instead. It would get very verbose otherwise to go to individual exceptions as there are a potential of 15 different exceptions possible.

File: docs/apache-airflow-providers-opensearch/operators/opensearch.rst
Comment (cjames23) [reply docs/apache-airflow-providers-opensearch/operators/opensearch.rst]: Fixed.

File: docs/apache-airflow-providers-opensearch/connections/opensearch.rst
Comment (cjames23) [reply docs/apache-airflow-providers-opensearch/connections/opensearch.rst]: Added what the default is set to.

File: airflow/providers/opensearch/hooks/opensearch.py
Comment (cjames23) [reply airflow/providers/opensearch/hooks/opensearch.py]: Good call out. Pushed change for it to be a cached_property

File: tests/providers/opensearch/conftest.py
Comment (ferruzzi) [inline tests/providers/opensearch/conftest.py]: I know it doesn't actually matter, but typo in `usser`

File: tests/providers/opensearch/conftest.py
Comment (ferruzzi) [inline tests/providers/opensearch/conftest.py:29]: I haven't seen monkeypatch used this way, I really like that.

File: tests/providers/opensearch/hooks/test_opensearch.py
Comment (ferruzzi) [inline tests/providers/opensearch/hooks/test_opensearch.py:25]: Does this actually work if you only run the tests in this file?   I wold have expected to need to import the mock_hook fixture

Comment (ferruzzi) [reply tests/providers/opensearch/hooks/test_opensearch.py:25]: and the fixture doesn't have autouse=True....   I wouldn't expect this to work as written

File: tests/providers/opensearch/hooks/test_opensearch.py
Comment (ferruzzi) [inline tests/providers/opensearch/hooks/test_opensearch.py]: [Here and elsewhere]  Functionally identical but perhaps a little easier to read if you split the hook out into its own assignment?  Not sure, I'll leave it up to you

```suggestion
        hook = OpenSearchHook(open_search_conn_id="opensearch_default",  log_query=True)
        result = hook.search(
            index_name="testIndex", 
            query={"size": 1, "query": {"multi_match": {"query": "test", "fields": ["testField"]}}},
        )
```

File: tests/providers/opensearch/hooks/test_opensearch.py
Comment (cjames23) [reply tests/providers/opensearch/hooks/test_opensearch.py:25]: Yes this works. See https://docs.pytest.org/en/6.2.x/fixture.html#scope-sharing-fixtures-across-classes-modules-packages-or-session

File: tests/providers/opensearch/operators/test_opensearch.py
Comment (ferruzzi) [inline tests/providers/opensearch/operators/test_opensearch.py]: Nitpick, and doesn't REALLY matter since it's a unit test, but this would parse out to "unit_teststest_schedule_dag_once", note the second "word" gets smooshed.

File: tests/providers/opensearch/operators/test_opensearch.py
Comment (ferruzzi) [inline tests/providers/opensearch/operators/test_opensearch.py]: Consider renaming MOCK_SEARCH_RETURN to EXPECTED_SEARCH_RETURN perhaps?

File: tests/providers/opensearch/operators/test_opensearch.py
Comment (ferruzzi) [inline tests/providers/opensearch/operators/test_opensearch.py]: You might consider moving the dag creation into a fixture since all these tests are repeating that block.  In addition, as a suggestion based on personal preference, I'd also inline the args.  There doesn't appear to be any reason to have them stored as a variable here.

File: tests/providers/opensearch/hooks/test_opensearch.py
Comment (cjames23) [reply tests/providers/opensearch/hooks/test_opensearch.py:25]: autouse=True would mean that tests do not have to explicitly request the fixture by having it defined as a parameter for the test function. I am just being explicit by requesting the fixture but I could switch it to autouse.

**Review by ferruzzi — COMMENTED** [review]
I left a handful of suggestions, but most are just my own personal preference and I won't block for that.  Approved as-is, but have a look and consider the suggestions.

The only potentially blocking comment is around the mock_hook fixture use in `tests/providers/opensearch/hooks/test_opensearch.py`

File: tests/providers/opensearch/hooks/test_opensearch.py
Comment (ferruzzi) [reply tests/providers/opensearch/hooks/test_opensearch.py:25]: If it works, it works.   I must have my scopes mixed up.  My mistake.

Comment (cjames23) [reply tests/providers/opensearch/hooks/test_opensearch.py:25]: Here is a output in breeze running just a single test. 
```
root@a8147a85fe4a:/opt/airflow/tests/providers/opensearch/hooks# pytest test_opensearch.py -k "test_hook_search"
========================================= test session starts =========================================
platform linux -- Python 3.10.13, pytest-7.4.2, pluggy-1.3.0 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /opt/airflow
configfile: pyproject.toml
plugins: httpx-0.21.3, mock-3.11.1, xdist-3.3.1, capture-warnings-0.0.4, instafail-0.5.0, anyio-4.0.0, rerunfailures-12.0, timeouts-1.2.1, asyncio-0.21.1, cov-4.1.0, time-machine-2.13.0, requests-mock-1.11.0
asyncio: mode=strict
setup timeout: 0.0s, execution timeout: 0.0s, teardown timeout: 0.0s
collected 2 items / 1 deselected / 1 selected                                                         

test_opensearch.py::TestOpenSearchHook::test_hook_search PASSED                                 [100%]

=================================== 1 passed, 1 deselected in 0.43s ===================================
```

File: tests/providers/opensearch/conftest.py
Comment (cjames23) [reply tests/providers/opensearch/conftest.py]: Fixed in latest push.

File: tests/providers/opensearch/operators/test_opensearch.py
Comment (cjames23) [reply tests/providers/opensearch/operators/test_opensearch.py]: Fixed in latest push.

File: tests/providers/opensearch/operators/test_opensearch.py
Comment (cjames23) [reply tests/providers/opensearch/operators/test_opensearch.py]: Renamed in latest push.

File: tests/providers/opensearch/operators/test_opensearch.py
Comment (cjames23) [reply tests/providers/opensearch/operators/test_opensearch.py]: Good callout. Moved this to a fixture in the latest push.

**Comment by ferruzzi** [comment]
I just noticed in the commits tab there are a couple of "chime notifier" commit messages.   Those should likely not be in the PR?
image

**Comment by cjames23** [comment]
@ferruzzi those should not have any impact. I think they showed up because I didn’t pay attention when creating this new branch as my local was still on the chime notifier branch. I can try to cherry pick them out if there is a blocking concern but all of that code was already merged.

**Comment by eladkal** [comment]
> @ferruzzi those should not have any impact. I think they showed up because I didn’t pay attention when creating this new branch as my local was still on the chime notifier branch. I can try to cherry pick them out if there is a blocking concern but all of that code was already merged.

I think would be easier to just squash all commits

File: airflow/providers/opensearch/hooks/opensearch.py
Comment (vincbeck) [inline airflow/providers/opensearch/hooks/opensearch.py]: nit: since it is a property, I would just call it `client`
```suggestion
    def client(self) -> OpenSearch:
```

**Comment by ferruzzi** [comment]
> I think would be easier to just squash all commits

Yeah, given that there is no impact, I'll agree there.

**Comment by cjames23** [comment]
Lazy consensus has been reached on adding this new provider from this PR.

File: airflow/providers/opensearch/hooks/opensearch.py
Comment (cjames23) [reply airflow/providers/opensearch/hooks/opensearch.py]: There was some miscommunication via the thread on the branches showing up from creating this branch not from mainline. So I tried to be proactive and drop those commits from this branch which ideally I don’t want to push that change unless it was absolutely necessary. I can clean this up if it is blocking, the suggestion you have can’t be applied given the references to this method in other places in the code.

**Comment by eladkal** [comment]
Just to clarify - This PR suppose to be only open search provider related code and https://github.com/apache/airflow/pull/34693 should be a followup with additional integrations for open search with other providers right?

Asking because if so then this PR should not modify Amazon provider code

**Comment by ferruzzi** [comment]
Yeah, I saw that, but the only changes in the Amazon package were some whitespace and punctuation in a docstring.   Better done separate, but I didn't think it was worth blocking over it.

**Comment by eladkal** [comment]
> Yeah, I saw that, but the only changes in the Amazon package were some whitespace and punctuation in a docstring. Better done separate, but I didn't think it was worth blocking over it.

Not a blocker but since rebase and resolving conflicts is required it can be done along with it

**Review by eladkal — APPROVED** [review]

**Review by vincbeck — APPROVED** [review]

File: airflow/providers/opensearch/hooks/opensearch.py
Comment (cjames23) [reply airflow/providers/opensearch/hooks/opensearch.py]: I had to push changes for breeze commands so I am pushing this change now.

**Comment by cjames23** [comment]
@eladkal any advice on the failure that is happening with the CI image build step and the static checks? I ran breeze static checks against all files locally and do not receive the same failure.

**Comment by cjames23** [comment]
> Yeah, I saw that, but the only changes in the Amazon package were some whitespace and punctuation in a docstring. Better done separate, but I didn't think it was worth blocking over it. 

Those changes were because pre-commit checks failed locally for those things after a merge pull.

**Comment by ferruzzi** [comment]
What an odd static check to fail, I can't see  how it could be related.  I'm going to bump it and see if it was just flake.   If it fails again, we'll have to sort it out.

**Comment by cjames23** [comment]
If the checks fail after this merge from main do you want me to include any fixes here for the breeze mypy issue? I think it just needs to be changed from ignore[misc] to ignore[attr-defined]. Not sure about the unit tests that are failing though from core.

**Comment by eladkal** [comment]
> If the checks fail after this merge from main do you want me to include any fixes here for the breeze mypy issue? I think it just needs to be changed from ignore[misc] to ignore[attr-defined]. Not sure about the unit tests that are failing though from core.

We have identified the issue
https://github.com/apache/airflow/pull/34941 should fix it

**Comment by potiuk** [comment]
Fixed https://github.com/apache/airflow/pull/34941 and few other issues that caused broken main. Rebasing/resolving conflicts (generally regenerating breeze images with. `breeze setup regenerate-images` should fix it.

**Review by eladkal — APPROVED** [review]

**Comment by eladkal** [comment]
🎉 🎉 🎉 🎉 🎉 🎉

**Comment by ferruzzi** [comment]
Nice, congrats!

**Comment by pateash** [comment]
Hi @cjames23,
Just a quick question regarding this one.
as we know OpenSearch is just a fork of Elasticsearch  which happened couple of years ago,
Can we not use Elastic search operators for connecting to OpenSearch,
 and if these hooks and operators are very similar to elastic search operators.

**Pull request merged**
